File: gitaccess.py
#! /usr/bin/env python

import os
import re
import sys
from pymongo import Connection

ssh_command = os.environ['SSH_ORIGINAL_COMMAND']
user = sys.argv[1]

m = re.search("^(git-receive-pack|git-upload-pack) '(\w+)\.git'$", ssh_command)
if m == None:
    exit(1)

command = m.group(1)
repoName = m.group(2)

connection = Connection('localhost', 27017)

db = connection.gitbox
repoCollection = db["Repository"]

repo = repoCollection.find_one({"name" : repoName, "writeUsers" : user})
if repo == None and command == "git-upload-pack":
    repo = repoCollection.find_one({"name" : repoName, "readUsers" : user})

if repo != None:
    arg = "%s 'repo/%s.git'" %(command, repoName)
    os.execv("/usr/bin/git-shell", ["/usr/bin/git-shell","-c", arg])

File: gitaccess.rb
#! /usr/bin/env ruby

require 'rubygems'  # not necessary for Ruby 1.9
require 'mongo'

#git-receive-pack 'toto.git'


user  = ARGV[0]

ssh_command = ENV['SSH_ORIGINAL_COMMAND']
#puts ssh_command

if ssh_command =~ /^(git-receive-pack|git-upload-pack) '(\w+)\.git'$/
  command = $1
  repoName = $2
else
  exit(1)
end
  
db = Mongo::Connection.new.db("gitbox")
coll = db["Repository"]
repository = coll.find({"name" => repoName, "writeUsers" => user}).first()
if repository == nil and command == "git-upload-pack"
  repository = coll.find({"name" => repoName, "readUsers" => user}).first()
end

if repository != nil
  exec("git-shell -c \"#{command} 'repo/#{repository["owner"]}/#{repoName}.git'\"")
end